Libraries/LibWeb/CredentialManagement/CredentialsContainer.cpp

```diff
@@ -17,25 +20,476 @@ GC::Ref<CredentialsContainer> CredentialsContainer::create(JS::Realm& realm)
+        if (options.mediation == Bindings::CredentialMediationRequirement::Silent) {
+            WebIDL::resolve_promise(realm(), *promise, JS::js_null());
+            return;
+        }
+
+        // 5. Let result be the result of asking the user to choose a Credential, given options and credentials.
+        auto result = ask_the_user_to_choose_a_credential(options, credentials);
+
+        // 6. If result is an interface object:
+        if (result.has<CredentialInterface*>()) {
+            // 1. Set result to the result of executing result’s [[DiscoverFromExternalSource]](origin, options, sameOriginWithAncestors),
+            //    given origin, options, and sameOriginWithAncestors.
+            auto maybe_result = result.get<CredentialInterface*>()->discover_from_external_source(realm(), origin, options, same_origin_with_ancestors);
+            // If that threw an exception:
+            if (maybe_result.is_error()) {
+                // 1. Let e be the thrown exception.
+                auto e = maybe_result.error_value();
+                // 2. Queue a task on global’s DOM manipulation task source to run the following substeps:
+                queue_global_task(HTML::Task::Source::DOMManipulation, settings.global_object(), GC::create_function(realm().heap(), [&] {
+                    HTML::TemporaryExecutionContext execution_context { realm(), HTML::TemporaryExecutionContext::CallbacksEnabled::Yes };
+
+                    // 1. Reject p with e.
+                    WebIDL::reject_promise(realm(), *promise, e);
+                }));
+                // 3. Terminate these substeps.
+                return;
+            }
+        }
+
+        // 7. Assert: result is null, or a Credential.
+        VERIFY(result.has<Empty>() || result.has<GC::Ref<Credential>>());
+
+        // 8. If result is a Credential, resolve p with result.
+        if (result.has<GC::Ref<Credential>>()) {
+            WebIDL::resolve_promise(realm(), *promise, result.get<GC::Ref<Credential>>());
+            return;
+        }
+
+        // 9. If result is null and options.mediation is not conditional, resolve p with result.
+        if (result.has<Empty>() && options.mediation != Bindings::CredentialMediationRequirement::Conditional)
+            WebIDL::resolve_promise(realm(), *promise, JS::js_null());
+    }));
+
+    // 14. React to p:
+    auto on_completion = GC::create_function(realm().heap(), [&settings, interfaces = move(interfaces)](JS::Value) -> WebIDL::ExceptionOr<JS::Value> {
+        // 1. For each interface in interfaces:
+        for (auto const& interface : interfaces) {
+            // 1. Remove interface’s [[type]] from settings’ active credential types.
+            settings.active_credential_types().remove_first_matching([&](auto& v) { return v == interface->type(); });
+        }
+
+        return JS::js_undefined();
+    });
+    WebIDL::react_to_promise(*promise, on_completion, on_completion);
```

Comment:
The specification of `react` in WebIDL makes me question whether the CredentialManagement spec really intended that the steps here should be performed if the promise is rejected.

Reply:
Oh `react`'s prose seems pretty unclear, but I think this is correct.

> To react to a Promise<T> promise, given one or two sets of steps to perform, covering when the promise is fulfilled, rejected, or both

I'm reading this as: if one callback was given, use it for both fulfillment and rejection. Which would mean our implementation is bugged.

Reply:
> if one callback was given, use it for both fulfillment and rejection

Yes that's how I read it too. And it also makes sense in this context as you want to `Remove interface’s [[type]] from settings’ active credential types.` for both success and failure.